### DISCUSSION TITLE
Vectorised instructions for S2?

### DISCUSSION THREAD

**Discussion by marklit** [klauspost/compress#732] [Q&A] [answered]
I've noticed S2 uses 47 unique opcodes in `encodeblock_amd64.s` but none are vectorised. Are there any plans to support AVX2 or Intel's other vectorisation opcode sets in the future?

**Comment by klauspost** [comment, marked as answer]
I have not found any useful SIMD instructions for compression.

There is a single interesting [BMI instruction](https://github.com/klauspost/compress/blob/master/s2/_generate/gen.go#L2624-L2630) which is 1 cycle faster on AMD Zen1-3. It is enabled at compile time by setting [GOAMD64=v3](https://github.com/golang/go/issues/45453) env var.

Didn't see significant enough speedup to duplicate all code for it. On most Intel CPUs and Zen4 they are equal anyway.

**Reply by marklit** [reply]
Okay, thank you.

**Reply by alexey-milovidov** [reply]
But vector instructions are useful for decompression.
See https://github.com/ClickHouse/ClickHouse/blob/master/src/Compression/LZ4_decompress_faster.cpp in ClickHouse.

**Reply by klauspost** [reply]
The [current code](https://github.com/klauspost/compress/blob/master/s2/decode_amd64.s#L431)  expands writes to 8 bytes, even with small offsets.

The case for calling "verySlowForwardCopy" will be the same, since your vector writes needs to respect the end of the destination.

You are of course welcome to send a PR.
